beacon_node/beacon_chain/src/beacon_chain.rs

```diff
@@ -3673,69 +3674,337 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
         &self,
-        finalized_execution_block_hash: ExecutionBlockHash,
-        head_block_root: Hash256,
-        head_execution_block_hash: ExecutionBlockHash,
+        current_slot: Slot,
     ) -> Result<(), Error> {
         let execution_layer = self
             .execution_layer
             .as_ref()
             .ok_or(Error::ExecutionLayerMissing)?;
 
         execution_layer
-            .block_on_generic(|_| {
-                self.update_execution_engine_forkchoice_async(
-                    finalized_execution_block_hash,
-                    head_block_root,
-                    head_execution_block_hash,
-                )
-            })
+            .block_on_generic(|_| self.update_execution_engine_forkchoice_async(current_slot))
             .map_err(Error::ForkchoiceUpdate)?
     }
 
     pub async fn update_execution_engine_forkchoice_async(
         &self,
-        finalized_execution_block_hash: ExecutionBlockHash,
-        head_block_root: Hash256,
-        head_execution_block_hash: ExecutionBlockHash,
+        current_slot: Slot,
     ) -> Result<(), Error> {
+        let execution_layer = self
+            .execution_layer
+            .as_ref()
+            .ok_or(Error::ExecutionLayerMissing)?;
+
+        // Take the global lock for updating the execution engine fork choice.
+        //
+        // Whilst holding this lock we must:
+        //
+        // 1. Read the canonical head.
+        // 2. Issue a forkchoiceUpdated call to the execution engine.
+        //
+        // This will allow us to ensure that we provide the execution layer with an *ordered* view
+        // of the head. I.e., we will never communicate a past head after communicating a later
+        // one.
+        //
+        // There are two "deadlock warnings" in this function. The downside of this nice ordering
+        // is the potential for deadlock. I would advise against any other use of
+        // `execution_engine_forkchoice_lock` apart from the one here.
+        let forkchoice_lock = execution_layer.execution_engine_forkchoice_lock().await;
+
+        // Deadlock warning:
+        //
+        // We are taking the `self.canonical_head` lock whilst holding the `forkchoice_lock`. This
+        // is intentional, since it allows us to ensure a consistent ordering of messages to the
+        // execution layer.
+        let head = self.head_info()?;
+        let head_block_root = head.block_root;
+        let head_execution_block_hash = if let Some(hash) = head.execution_payload_block_hash {
+            hash
+        } else {
+            // Don't send fork choice updates to the execution layer before the transition block.
+            return Ok(());
+        };
+
+        let finalized_root = if head.finalized_checkpoint.root == Hash256::zero() {
+            // De-alias `0x00..00` to the genesis block root.
+            self.genesis_block_root
+        } else {
+            head.finalized_checkpoint.root
+        };
```

Comment:
Ah yes, good question!

We didn't need to do it in the last PR since we were getting the root from fork choice itself. Fork choice does the de-aliasing for us, so no need there. Here we're getting the root from the `BeaconState` which doesn't do the de-aliasing. I didn't notice this at first and found it in a test!